From f2ee45827bc2060a2e98d233288cc3287dbd94bf Mon Sep 17 00:00:00 2001
From: Hao Zheng <haoz@marvell.com>
Date: Thu, 13 Jun 2019 21:41:25 -0700
Subject: [PATCH 0206/1921] octeontx2-af: reorganize NPC parser profile for
 tunnel protocols

To better support rss for tunnelled packets, udp based tunnel
protocols such as vxlan, vxlan-gpe, geneve and gtpu are now
captured into a separate layer E. Consequently, the inner
packet headers are pushed one layer down to LF, LG, and LH
accordingly. Some minor reorganization of some other tunnel
protocols are also performed to make the whole parser profile
uniform.

Change-Id: I06fe5f3aae81c49e18b0b16643a790d4fb6e021d
Signed-off-by: Hao Zheng <haoz@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/10911
Reviewed-by: Kiran Kumar Kokkilagadda <Kiran.Kokkilagadda@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/af/npc.h   |   68 +-
 .../marvell/octeontx2/af/npc_profile.h        | 2522 ++++++++---------
 .../ethernet/marvell/octeontx2/af/rvu_nix.c   |   72 +-
 .../ethernet/marvell/octeontx2/af/rvu_npc.c   |    2 +-
 4 files changed, 1286 insertions(+), 1378 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/npc.h b/drivers/net/ethernet/marvell/octeontx2/af/npc.h
index f4fff4c8afa6..ed450a6e81d7 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/npc.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/npc.h
@@ -58,63 +58,61 @@ enum npc_kpu_lc_ltype {
 /* Don't modify Ltypes upto SCTP, otherwise it will
  * effect flow tag calculation and thus RSS.
  */
-/* Dont change values of UDP_VXLAN(12) and UDP_GENEVE(13), otherwise it
- * will effect flow tag calculation and thus RSS
- */
 enum npc_kpu_ld_ltype {
 	NPC_LT_LD_TCP = 1,
 	NPC_LT_LD_UDP,
 	NPC_LT_LD_ICMP,
 	NPC_LT_LD_SCTP,
-	NPC_LT_LD_IGMP,
 	NPC_LT_LD_ICMP6,
-	NPC_LT_LD_TU_MPLS,
+	NPC_LT_LD_IGMP = 8,
 	NPC_LT_LD_ESP,
 	NPC_LT_LD_AH,
 	NPC_LT_LD_GRE,
 	NPC_LT_LD_NVGRE,
-	NPC_LT_LD_UDP_VXLAN,
-	NPC_LT_LD_UDP_GENEVE,
-	NPC_LT_LD_UDP_GTPU,
+	NPC_LT_LD_NSH,
+	NPC_LT_LD_TU_MPLS_IN_NSH,
+	NPC_LT_LD_TU_MPLS_IN_IP,
 };
 
 enum npc_kpu_le_ltype {
-	NPC_LT_LE_TU_ETHER = 1,
-	NPC_LT_LE_TU_PPP,
-	NPC_LT_LE_TU_MPLS_IN_NSH,
-	NPC_LT_LE_TU_3RD_NSH,
+	NPC_LT_LE_VXLAN = 1,
+	NPC_LT_LE_GENEVE,
+	NPC_LT_LE_GTPU = 4,
+	NPC_LT_LE_VXLANGPE,
+	NPC_LT_LE_GTPC,
+	NPC_LT_LE_NSH,
+	NPC_LT_LE_TU_MPLS_IN_GRE,
+	NPC_LT_LE_TU_NSH_IN_GRE,
 };
 
 enum npc_kpu_lf_ltype {
-	NPC_LT_LF_TU_IP = 1,
-	NPC_LT_LF_TU_IP6,
-	NPC_LT_LF_TU_ARP,
-	NPC_LT_LF_TU_MPLS_IP,
-	NPC_LT_LF_TU_MPLS_IP6,
-	NPC_LT_LF_TU_MPLS_ETHER,
+	NPC_LT_LF_TU_ETHER = 1,
+	NPC_LT_LF_TU_PPP,
+	NPC_LT_LF_TU_MPLS_IN_VXLANGPE,
+	NPC_LT_LF_TU_NSH_IN_VXLANGPE,
+	NPC_LT_LF_TU_MPLS_IN_NSH,
+	NPC_LT_LF_TU_3RD_NSH,
 };
 
-/* Don't modify Ltypes upto SCTP, otherwise it will
- * effect flow tag calculation and thus RSS.
- */
 enum npc_kpu_lg_ltype {
-	NPC_LT_LG_TU_TCP = 1,
-	NPC_LT_LG_TU_UDP,
-	NPC_LT_LG_TU_ICMP,
-	NPC_LT_LG_TU_SCTP,
-	NPC_LT_LG_TU_IGMP,
-	NPC_LT_LG_TU_ICMP6,
-	NPC_LT_LG_TU_ESP,
-	NPC_LT_LG_TU_AH,
+	NPC_LT_LG_TU_IP = 1,
+	NPC_LT_LG_TU_IP6,
+	NPC_LT_LG_TU_ARP,
+	NPC_LT_LG_TU_ETHER_IN_NSH,
 };
 
+/* Don't modify Ltypes upto SCTP, otherwise it will
+ * effect flow tag calculation and thus RSS.
+ */
 enum npc_kpu_lh_ltype {
-	NPC_LT_LH_TCP_DATA = 1,
-	NPC_LT_LH_HTTP_DATA,
-	NPC_LT_LH_HTTPS_DATA,
-	NPC_LT_LH_PPTP_DATA,
-	NPC_LT_LH_UDP_DATA,
-	NPC_LT_LH_UDP_PTP,
+	NPC_LT_LH_TU_TCP = 1,
+	NPC_LT_LH_TU_UDP,
+	NPC_LT_LH_TU_ICMP,
+	NPC_LT_LH_TU_SCTP,
+	NPC_LT_LH_TU_ICMP6,
+	NPC_LT_LH_TU_IGMP = 8,
+	NPC_LT_LH_TU_ESP,
+	NPC_LT_LH_TU_AH,
 };
 
 struct npc_kpu_profile_cam {
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h b/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
index 6fb2d34abf09..d8a6d83e5f52 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
@@ -181,16 +181,25 @@ enum npc_kpu_parser_state {
 	NPC_S_KPU8_GRE,
 	NPC_S_KPU8_ESP,
 	NPC_S_KPU8_AH,
-	NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN,
-	NPC_S_KPU9_TU_MPLS,
-	NPC_S_KPU9_TU_NSH,
+	NPC_S_KPU9_TU_MPLS_IN_GRE,
+	NPC_S_KPU9_TU_MPLS_IN_NSH,
+	NPC_S_KPU9_TU_MPLS_IN_IP,
+	NPC_S_KPU9_TU_NSH_IN_GRE,
+	NPC_S_KPU9_VXLAN,
+	NPC_S_KPU9_VXLANGPE,
+	NPC_S_KPU9_GENEVE,
+	NPC_S_KPU9_GTPC,
+	NPC_S_KPU9_GTPU,
+	NPC_S_KPU10_TU_MPLS_IN_VXLANGPE,
 	NPC_S_KPU10_TU_MPLS_PL,
 	NPC_S_KPU10_TU_MPLS,
-	NPC_S_KPU10_TU_NSH,
+	NPC_S_KPU10_TU_NSH_IN_VXLANGPE,
 	NPC_S_KPU11_TU_ETHER,
 	NPC_S_KPU11_TU_PPP,
 	NPC_S_KPU11_TU_MPLS_IN_NSH,
-	NPC_S_KPU11_TU_3RD_NSH,
+	NPC_S_KPU11_TU_MPLS_PL,
+	NPC_S_KPU11_TU_MPLS,
+	NPC_S_KPU11_TU_ETHER_IN_NSH,
 	NPC_S_KPU12_TU_IP,
 	NPC_S_KPU12_TU_IP6,
 	NPC_S_KPU12_TU_ARP,
@@ -267,26 +276,11 @@ enum npc_kpu_lc_lflag {
 	NPC_F_LC_L_EXT_SHIM6,
 };
 
-enum npc_kpu_ld_uflag {
-	NPC_F_LD_U_MPLS_IN_GRE = 0x80,
-	NPC_F_LD_U_NSH_IN_GRE = 0x40,
-};
 enum npc_kpu_ld_lflag {
 	NPC_F_LD_L_TCP_UNK_PORT = 1,
 	NPC_F_LD_L_TCP_HAS_OPTIONS,
 	NPC_F_LD_L_TCP_UNK_PORT_HAS_OPTIONS,
-	NPC_F_LD_L_UDP_VXLAN_NOVNI,
-	NPC_F_LD_L_UDP_VXLANGPE,
-	NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
-	NPC_F_LD_L_UDP_VXLANGPE_UNK,
-	NPC_F_LD_L_UDP_VXLANGPE_NONP,
-	NPC_F_LD_L_UDP_GTP_GTPC,
-	NPC_F_LD_L_UDP_GTP_GTPU_G_PDU,
-	NPC_F_LD_L_UDP_GTP_GTPU_UNK,
 	NPC_F_LD_L_UDP_UNK_PORT,
-	NPC_F_LD_L_UDP_GENEVE_OAM,
-	NPC_F_LD_L_UDP_GENEVE_CRI_OPT,
-	NPC_F_LD_L_UDP_GENEVE_OAM_CRI_OPT,
 	NPC_F_LD_L_GRE_NVGRE,
 	NPC_F_LD_L_GRE_HAS_SRE,
 	NPC_F_LD_L_GRE_HAS_CSUM,
@@ -308,33 +302,47 @@ enum npc_kpu_ld_lflag {
 	NPC_F_LD_L_MPLS_2_LABELS,
 };
 
-enum npc_kpu_le_uflag {
-	NPC_F_LE_U_UNK_ETYPE = 0x10,
-	NPC_F_LE_U_HAS_TAG = 0x20,
-};
 enum npc_kpu_le_lflag {
-	NPC_F_LE_L_WITH_CTAG = 1,
-	NPC_F_LE_L_WITH_STAG_CTAG,
-	NPC_F_LE_L_WITH_STAG,
-	NPC_F_LE_L_WITH_QINQ_CTAG,
-	NPC_F_LE_L_WITH_QINQ,
+	NPC_F_LE_L_VXLAN_NOVNI,
+	NPC_F_LE_L_VXLANGPE_NOVNI,
+	NPC_F_LE_L_VXLANGPE_UNK,
+	NPC_F_LE_L_VXLANGPE_NONP,
+	NPC_F_LE_L_GENEVE_OAM,
+	NPC_F_LE_L_GENEVE_CRI_OPT,
+	NPC_F_LE_L_GENEVE_OAM_CRI_OPT,
+	NPC_F_LE_L_GTPU_G_PDU,
+	NPC_F_LE_L_GTPU_UNK,
 };
 
 enum npc_kpu_lf_uflag {
-	NPC_F_LF_U_UNK_IP_PROTO = 0x10,
-	NPC_F_LF_U_IP_HAS_OPTIONS = 0x20,
-	NPC_F_LF_U_IP6_HAS_EXT = 0x40,
+	NPC_F_LF_U_UNK_ETYPE = 0x10,
+	NPC_F_LF_U_HAS_TAG = 0x20,
+};
+
+enum npc_kpu_lf_lflag {
+	NPC_F_LF_L_WITH_CTAG = 1,
+	NPC_F_LF_L_WITH_STAG_CTAG,
+	NPC_F_LF_L_WITH_STAG,
+	NPC_F_LF_L_WITH_QINQ_CTAG,
+	NPC_F_LF_L_WITH_QINQ,
 };
 
 enum npc_kpu_lg_uflag {
-	NPC_F_LG_U_TCP_HAS_OPTIONS = 0x80,
+	NPC_F_LG_U_UNK_IP_PROTO = 0x10,
+	NPC_F_LG_U_IP_HAS_OPTIONS = 0x20,
+	NPC_F_LG_U_IP6_HAS_EXT = 0x40,
 };
-enum npc_kpu_lg_lflag {
-	NPC_F_LG_L_TCP_HTTP = 1,
-	NPC_F_LG_L_TCP_HTTPS,
-	NPC_F_LG_L_TCP_PPTP,
-	NPC_F_LG_L_TCP_UNK_PORT,
-	NPC_F_LG_L_UDP_UNK_PORT,
+
+enum npc_kpu_lh_uflag {
+	NPC_F_LH_U_TCP_HAS_OPTIONS = 0x80,
+};
+
+enum npc_kpu_lh_lflag {
+	NPC_F_LH_L_TCP_HTTP = 1,
+	NPC_F_LH_L_TCP_HTTPS,
+	NPC_F_LH_L_TCP_PPTP,
+	NPC_F_LH_L_TCP_UNK_PORT,
+	NPC_F_LH_L_UDP_UNK_PORT,
 };
 
 enum npc_kpu_err_code {
@@ -345,8 +353,10 @@ enum npc_kpu_err_code {
 	NPC_EC_L2_K2,
 	NPC_EC_L2_K3,
 	NPC_EC_L2_K3_ETYPE_UNK,
-	NPC_EC_L2_MPLS_2MANY,
 	NPC_EC_L2_K4,
+	NPC_EC_MPLS_2MANY,
+	NPC_EC_MPLS_UNK,
+	NPC_EC_NSH_UNK,
 	NPC_EC_IP_TTL_0,
 	NPC_EC_IP_FRAG_OFFSET_1,
 	NPC_EC_IP_VER,
@@ -1628,87 +1638,6 @@ static struct npc_kpu_profile_cam kpu2_cam_entries[] = {
 		0x0000,
 		0x0000,
 	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_IP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_IP6,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_ARP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_RARP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_PTP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_FCOE,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_MPLSU,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_MPLSM,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_SBTAG, 0xff,
-		NPC_ETYPE_NSH,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
 	{
 		NPC_S_KPU2_SBTAG, 0xff,
 		0x0000,
@@ -1826,87 +1755,6 @@ static struct npc_kpu_profile_cam kpu2_cam_entries[] = {
 		0x0000,
 		0x0000,
 	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_IP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_IP6,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_ARP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_RARP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_PTP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_FCOE,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_MPLSU,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_MPLSM,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU2_QINQ, 0xff,
-		NPC_ETYPE_NSH,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
 	{
 		NPC_S_KPU2_QINQ, 0xff,
 		0x0000,
@@ -3526,7 +3374,7 @@ static struct npc_kpu_profile_cam kpu4_cam_entries[] = {
 	},
 	{
 		NPC_S_KPU4_NSH, 0xff,
-		NPC_NSH_NP_NSH,
+		NPC_NSH_NP_MPLS,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -3535,8 +3383,8 @@ static struct npc_kpu_profile_cam kpu4_cam_entries[] = {
 	},
 	{
 		NPC_S_KPU4_NSH, 0xff,
-		NPC_NSH_NP_MPLS,
-		NPC_NSH_NP_MASK,
+		0x0000,
+		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
@@ -4112,51 +3960,6 @@ static struct npc_kpu_profile_cam kpu5_cam_entries[] = {
 		0x0000,
 		0x0000,
 	},
-	{
-		NPC_S_KPU5_NSH, 0xff,
-		NPC_NSH_NP_IP,
-		NPC_NSH_NP_MASK,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU5_NSH, 0xff,
-		NPC_NSH_NP_IP6,
-		NPC_NSH_NP_MASK,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU5_NSH, 0xff,
-		NPC_NSH_NP_ETH,
-		NPC_NSH_NP_MASK,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU5_NSH, 0xff,
-		NPC_NSH_NP_NSH,
-		NPC_NSH_NP_MASK,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU5_NSH, 0xff,
-		NPC_NSH_NP_MPLS,
-		NPC_NSH_NP_MASK,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
 	{
 		NPC_S_NA, 0X00,
 		0x0000,
@@ -4863,23 +4666,23 @@ static struct npc_kpu_profile_cam kpu8_cam_entries[] = {
 		NPC_S_KPU8_UDP, 0xff,
 		NPC_UDP_PORT_VXLAN,
 		0xffff,
-		NPC_VXLAN_I,
-		NPC_VXLAN_I,
 		0x0000,
-		0xffff,
+		0x0000,
+		0x0000,
+		0x0000,
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLAN,
+		NPC_UDP_PORT_VXLANGPE,
 		0xffff,
 		0x0000,
-		0xffff,
 		0x0000,
-		0xffff,
+		0x0000,
+		0x0000,
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLAN,
+		NPC_UDP_PORT_GENEVE,
 		0xffff,
 		0x0000,
 		0x0000,
@@ -4888,259 +4691,34 @@ static struct npc_kpu_profile_cam kpu8_cam_entries[] = {
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
+		NPC_UDP_PORT_GTPC,
 		0xffff,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_IP,
-		NPC_VXLANGPE_NP_MASK,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
+		NPC_UDP_PORT_GTPU,
 		0xffff,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_IP6,
-		NPC_VXLANGPE_NP_MASK,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
+		NPC_UDP_PORT_PTP_E,
 		0xffff,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_ETH,
-		NPC_VXLANGPE_NP_MASK,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
 	},
 	{
 		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_NSH,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_MPLS,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_IP,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_IP6,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_ETH,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_NSH,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P|NPC_VXLANGPE_I,
-		NPC_VXLANGPE_NP_MPLS,
-		NPC_VXLANGPE_NP_MASK,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		NPC_VXLANGPE_P,
-		NPC_VXLANGPE_P,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_VXLANGPE,
-		0xffff,
-		0x0000,
-		NPC_VXLANGPE_P,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		0x0000,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_TRANS_ETH_BR,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_TRANS_ETH_BR,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_TRANS_ETH_BR,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_TRANS_ETH_BR,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		0x0000,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		0x0000,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP6,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP6,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP6,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GENEVE,
-		0xffff,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_GENEVE_F_OAM|NPC_GENEVE_F_CRI_OPT,
-		NPC_ETYPE_IP6,
-		0xffff,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GTPC,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GTPU,
-		0xffff,
-		NPC_GTP_PT_GTP|NPC_GTP_VER1|NPC_GTP_MT_G_PDU,
-		NPC_GTP_PT_MASK|NPC_GTP_VER_MASK|NPC_GTP_MT_MASK,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_GTPU,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_PTP_E,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU8_UDP, 0xff,
-		NPC_UDP_PORT_PTP_G,
+		NPC_UDP_PORT_PTP_G,
 		0xffff,
 		0x0000,
 		0x0000,
@@ -5682,7 +5260,7 @@ static struct npc_kpu_profile_cam kpu8_cam_entries[] = {
 
 static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 	{
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 0xff,
 		NPC_MPLS_S,
 		NPC_MPLS_S,
 		0x0000,
@@ -5691,7 +5269,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		NPC_MPLS_S,
@@ -5700,7 +5278,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		0x0000,
@@ -5709,7 +5287,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		0x0000,
@@ -5718,7 +5296,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_NSH, 0xff,
 		NPC_MPLS_S,
 		NPC_MPLS_S,
 		0x0000,
@@ -5727,7 +5305,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_NSH, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		NPC_MPLS_S,
@@ -5736,7 +5314,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_NSH, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		0x0000,
@@ -5745,7 +5323,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU9_TU_MPLS, 0xff,
+		NPC_S_KPU9_TU_MPLS_IN_NSH, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		0x0000,
@@ -5754,26 +5332,44 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU9_TU_NSH, 0xff,
-		NPC_NSH_NP_IP,
-		NPC_NSH_NP_MASK,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
 		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_NSH, 0xff,
-		NPC_NSH_NP_IP6,
-		NPC_NSH_NP_MASK,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 0xff,
+		0x0000,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU9_TU_MPLS_IN_IP, 0xff,
+		0x0000,
+		NPC_MPLS_S,
 		0x0000,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+	},
+	{
+		NPC_S_KPU9_TU_MPLS_IN_IP, 0xff,
 		0x0000,
+		NPC_MPLS_S,
 		0x0000,
+		NPC_MPLS_S,
 		0x0000,
+		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU9_TU_NSH, 0xff,
-		NPC_NSH_NP_ETH,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 0xff,
+		NPC_NSH_NP_IP,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -5781,8 +5377,8 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_NSH, 0xff,
-		NPC_NSH_NP_NSH,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 0xff,
+		NPC_NSH_NP_IP6,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -5790,8 +5386,8 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU9_TU_NSH, 0xff,
-		NPC_NSH_NP_MPLS,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 0xff,
+		NPC_NSH_NP_ETH,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -5799,7 +5395,7 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_NA, 0X00,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 0xff,
 		0x0000,
 		0x0000,
 		0x0000,
@@ -5807,47 +5403,326 @@ static struct npc_kpu_profile_cam kpu9_cam_entries[] = {
 		0x0000,
 		0x0000,
 	},
-};
-
-static struct npc_kpu_profile_cam kpu10_cam_entries[] = {
 	{
-		NPC_S_KPU10_TU_MPLS, 0xff,
-		NPC_MPLS_S,
-		NPC_MPLS_S,
-		NPC_IP_VER_4,
-		NPC_IP_VER_MASK,
+		NPC_S_KPU9_VXLAN, 0xff,
 		0x0000,
 		0x0000,
+		NPC_VXLAN_I,
+		NPC_VXLAN_I,
+		0x0000,
+		0xffff,
 	},
 	{
-		NPC_S_KPU10_TU_MPLS, 0xff,
-		NPC_MPLS_S,
-		NPC_MPLS_S,
-		NPC_IP_VER_6,
-		NPC_IP_VER_MASK,
+		NPC_S_KPU9_VXLAN, 0xff,
 		0x0000,
 		0x0000,
-	},
-	{
-		NPC_S_KPU10_TU_MPLS, 0xff,
-		NPC_MPLS_S,
-		NPC_MPLS_S,
 		0x0000,
 		0xffff,
 		0x0000,
-		0x0000,
+		0xffff,
 	},
 	{
-		NPC_S_KPU10_TU_MPLS, 0xff,
-		NPC_MPLS_S,
-		NPC_MPLS_S,
+		NPC_S_KPU9_VXLAN, 0xff,
+		0x0000,
+		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
 	},
 	{
-		NPC_S_KPU10_TU_MPLS, 0xff,
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_IP,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_IP6,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_ETH,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_NSH,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_MPLS,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_IP,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_IP6,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_ETH,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_NSH,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P | NPC_VXLANGPE_I,
+		NPC_VXLANGPE_NP_MPLS,
+		NPC_VXLANGPE_NP_MASK,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		NPC_VXLANGPE_P,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU9_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		NPC_VXLANGPE_P,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_TRANS_ETH_BR,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_TRANS_ETH_BR,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_TRANS_ETH_BR,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_TRANS_ETH_BR,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP6,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP6,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP6,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GENEVE, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_GENEVE_F_OAM | NPC_GENEVE_F_CRI_OPT,
+		NPC_ETYPE_IP6,
+		0xffff,
+	},
+	{
+		NPC_S_KPU9_GTPC, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU9_GTPU, 0xff,
+		0x0000,
+		0x0000,
+		NPC_GTP_PT_GTP | NPC_GTP_VER1 | NPC_GTP_MT_G_PDU,
+		NPC_GTP_PT_MASK | NPC_GTP_VER_MASK | NPC_GTP_MT_MASK,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU9_GTPU, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_NA, 0X00,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+};
+
+static struct npc_kpu_profile_cam kpu10_cam_entries[] = {
+	{
+		NPC_S_KPU10_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_IP_VER_4,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU10_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_IP_VER_6,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU10_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		0x0000,
+		0xffff,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU10_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU10_TU_MPLS, 0xff,
 		0x0000,
 		NPC_MPLS_S,
 		0x0000,
@@ -5892,26 +5767,44 @@ static struct npc_kpu_profile_cam kpu10_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU10_TU_NSH, 0xff,
-		NPC_NSH_NP_IP,
-		NPC_NSH_NP_MASK,
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
 		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
 	},
 	{
-		NPC_S_KPU10_TU_NSH, 0xff,
-		NPC_NSH_NP_IP6,
-		NPC_NSH_NP_MASK,
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 0xff,
 		0x0000,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
 		0x0000,
 		0x0000,
+	},
+	{
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 0xff,
+		0x0000,
+		NPC_MPLS_S,
 		0x0000,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
 	},
 	{
-		NPC_S_KPU10_TU_NSH, 0xff,
-		NPC_NSH_NP_ETH,
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 0xff,
+		0x0000,
+		NPC_MPLS_S,
+		0x0000,
+		NPC_MPLS_S,
+		0x0000,
+		NPC_MPLS_S,
+	},
+	{
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 0xff,
+		NPC_NSH_NP_IP,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -5919,8 +5812,8 @@ static struct npc_kpu_profile_cam kpu10_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU10_TU_NSH, 0xff,
-		NPC_NSH_NP_NSH,
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 0xff,
+		NPC_NSH_NP_IP6,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
@@ -5928,14 +5821,23 @@ static struct npc_kpu_profile_cam kpu10_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU10_TU_NSH, 0xff,
-		NPC_NSH_NP_MPLS,
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 0xff,
+		NPC_NSH_NP_ETH,
 		NPC_NSH_NP_MASK,
 		0x0000,
 		0x0000,
 		0x0000,
 		0x0000,
 	},
+	{
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 0xff,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
 	{
 		NPC_S_NA, 0X00,
 		0x0000,
@@ -6174,7 +6076,61 @@ static struct npc_kpu_profile_cam kpu11_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU11_TU_MPLS_IN_NSH, 0xff,
+		NPC_S_KPU11_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_IP_VER_4,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		NPC_IP_VER_6,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS, 0xff,
+		NPC_MPLS_S,
+		NPC_MPLS_S,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS, 0xff,
+		0x0000,
+		NPC_MPLS_S,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS_PL, 0xff,
+		NPC_IP_VER_4,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS_PL, 0xff,
+		NPC_IP_VER_6,
+		NPC_IP_VER_MASK,
+		0x0000,
+		0x0000,
+		0x0000,
+		0x0000,
+	},
+	{
+		NPC_S_KPU11_TU_MPLS_PL, 0xff,
 		0x0000,
 		0x0000,
 		0x0000,
@@ -6183,7 +6139,7 @@ static struct npc_kpu_profile_cam kpu11_cam_entries[] = {
 		0x0000,
 	},
 	{
-		NPC_S_KPU11_TU_3RD_NSH, 0xff,
+		NPC_S_KPU11_TU_ETHER_IN_NSH, 0xff,
 		0x0000,
 		0x0000,
 		0x0000,
@@ -7327,78 +7283,6 @@ static struct npc_kpu_profile_action kpu2_action_entries[] = {
 		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_ITAG_UNK,
 		0, 0, 0, 0,
 	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 2, 0,
-		NPC_S_KPU5_IP, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 2, 0,
-		NPC_S_KPU5_IP6, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_ARP, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_RARP, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_PTP, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_FCOE, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 1, 0,
-		NPC_S_KPU4_NSH, 6, 1,
-		NPC_LID_LB, NPC_LT_LB_STAG,
-		0,
-		0, 0, 0, 0,
-	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
@@ -7409,170 +7293,98 @@ static struct npc_kpu_profile_action kpu2_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 2, 0,
-		NPC_S_KPU5_IP, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 2, 0,
-		NPC_S_KPU5_IP6, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_ARP, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_RARP, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_PTP, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_FCOE, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 1, 0,
-		NPC_S_KPU4_NSH, 10, 1,
-		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU5_IP, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_CTAG_UNK,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 0, 0, 0,
-		NPC_S_KPU3_CTAG, 10, 1,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU5_IP6, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_QINQ_CTAG,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 0, 0, 0,
-		NPC_S_KPU3_QINQ, 10, 1,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU5_ARP, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		NPC_F_LB_U_MORE_TAG|NPC_F_LB_L_WITH_QINQ_QINQ,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 2, 0,
-		NPC_S_KPU5_IP, 6, 1,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU5_RARP, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 2, 0,
-		NPC_S_KPU5_IP6, 6, 1,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU5_PTP, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
-		NPC_S_KPU5_ARP, 6, 1,
+		NPC_S_KPU5_FCOE, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_RARP, 6, 1,
+		2, 6, 10, 1, 0,
+		NPC_S_KPU4_MPLS, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_PTP, 6, 1,
+		2, 6, 10, 1, 0,
+		NPC_S_KPU4_MPLS, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU5_FCOE, 6, 1,
+		2, 0, 0, 1, 0,
+		NPC_S_KPU4_NSH, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 6, 1,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_CTAG_UNK,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 1, 0,
-		NPC_S_KPU4_MPLS, 6, 1,
+		2, 6, 0, 0, 0,
+		NPC_S_KPU3_CTAG, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_QINQ_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 1, 0,
-		NPC_S_KPU4_NSH, 6, 1,
+		2, 6, 0, 0, 0,
+		NPC_S_KPU3_QINQ, 10, 1,
 		NPC_LID_LB, NPC_LT_LB_QINQ,
-		0,
+		NPC_F_LB_U_MORE_TAG | NPC_F_LB_L_WITH_QINQ_QINQ,
 		0, 0, 0, 0,
 	},
 	{
@@ -9015,19 +8827,19 @@ static struct npc_kpu_profile_action kpu4_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 0, 0,
-		NPC_S_KPU5_NSH, 0, 1,
+		0, 0, 0, 4, 0,
+		NPC_S_KPU9_TU_MPLS_IN_NSH, 0, 1,
 		NPC_LID_LC, NPC_LT_LC_NSH,
 		0,
 		1, 0x3f, 0, 2,
 	},
 	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 4, 0,
-		NPC_S_KPU9_TU_MPLS, 0, 1,
+		NPC_ERRLEV_LC, NPC_EC_NSH_UNK,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
 		NPC_LID_LC, NPC_LT_LC_NSH,
 		0,
-		1, 0x3f, 0, 2,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_LB, NPC_EC_L2_K4,
@@ -9139,7 +8951,7 @@ static struct npc_kpu_profile_action kpu5_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 3, 0,
-		NPC_S_KPU9_TU_MPLS, 20, 1,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 20, 1,
 		NPC_LID_LC, NPC_LT_LC_IP,
 		NPC_F_LC_L_MPLS_IN_IP,
 		0, 0, 0, 0,
@@ -9243,7 +9055,7 @@ static struct npc_kpu_profile_action kpu5_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 3, 0,
-		NPC_S_KPU9_TU_MPLS, 20, 1,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 20, 1,
 		NPC_LID_LC, NPC_LT_LC_IP_OPT,
 		NPC_F_LC_L_MPLS_IN_IP,
 		0, 0xf, 0, 2,
@@ -9371,7 +9183,7 @@ static struct npc_kpu_profile_action kpu5_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 3, 0,
-		NPC_S_KPU9_TU_MPLS, 40, 1,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 40, 1,
 		NPC_LID_LC, NPC_LT_LC_IP6,
 		NPC_F_LC_L_IP6_MPLS_IN_IP,
 		0, 0, 0, 0,
@@ -9497,7 +9309,7 @@ static struct npc_kpu_profile_action kpu5_action_entries[] = {
 		0, 0, 0, 0,
 	},
 	{
-		NPC_ERRLEV_LB, NPC_EC_L2_MPLS_2MANY,
+		NPC_ERRLEV_LB, NPC_EC_MPLS_2MANY,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
 		NPC_LID_LB, NPC_LT_NA,
@@ -9536,46 +9348,6 @@ static struct npc_kpu_profile_action kpu5_action_entries[] = {
 		0,
 		0, 0, 0, 0,
 	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 6, 0,
-		NPC_S_KPU12_TU_IP, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 6, 0,
-		NPC_S_KPU12_TU_IP6, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 5, 0,
-		NPC_S_KPU11_TU_ETHER, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 5, 0,
-		NPC_S_KPU11_TU_3RD_NSH, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 3, 0,
-		NPC_S_KPU9_TU_MPLS, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
-	},
 	{
 		NPC_ERRLEV_LC, NPC_EC_UNK,
 		0, 0, 0, 0, 1,
@@ -9670,7 +9442,7 @@ static struct npc_kpu_profile_action kpu6_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 2, 0,
-		NPC_S_KPU9_TU_MPLS, 8, 0,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 8, 0,
 		NPC_LID_LC, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
@@ -9758,7 +9530,7 @@ static struct npc_kpu_profile_action kpu6_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 2, 0,
-		NPC_S_KPU9_TU_MPLS, 8, 0,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 8, 0,
 		NPC_LID_LC, NPC_LT_NA,
 		0,
 		1, 0xff, 0, 3,
@@ -9862,7 +9634,7 @@ static struct npc_kpu_profile_action kpu6_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 2, 0,
-		NPC_S_KPU9_TU_MPLS, 8, 0,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 8, 0,
 		NPC_LID_LC, NPC_LT_NA,
 		0,
 		1, 0xff, 0, 3,
@@ -9977,7 +9749,7 @@ static struct npc_kpu_profile_action kpu7_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 1, 0,
-		NPC_S_KPU9_TU_MPLS, 8, 0,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 8, 0,
 		NPC_LID_LC, NPC_LT_NA,
 		0,
 		1, 0xff, 0, 3,
@@ -10065,7 +9837,7 @@ static struct npc_kpu_profile_action kpu7_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 1, 0,
-		NPC_S_KPU9_TU_MPLS, 8, 0,
+		NPC_S_KPU9_TU_MPLS_IN_IP, 8, 0,
 		NPC_LID_LC, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
@@ -10203,242 +9975,42 @@ static struct npc_kpu_profile_action kpu8_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_VXLAN,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_VXLAN,
-		NPC_F_LD_L_UDP_VXLAN_NOVNI,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_LD, NPC_EC_VXLAN,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
+		0, 0, 2, 0, 0,
+		NPC_S_KPU9_VXLAN, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 16, 1,
+		0, 0, 2, 0, 0,
+		NPC_S_KPU9_VXLANGPE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 16, 1,
+		0, 0, 2, 0, 0,
+		NPC_S_KPU9_GENEVE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NOVNI,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
+		0, 0, 2, 0, 0,
+		NPC_S_KPU9_GTPC, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_UNK,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
+		0, 0, 2, 0, 0,
+		NPC_S_KPU9_GTPU, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_VXLANGPE_NONP,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		0,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 2, 0,
-		NPC_S_KPU11_TU_ETHER, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		0,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
 		0,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 3, 0,
-		NPC_S_KPU12_TU_IP6, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GENEVE,
-		NPC_F_LD_L_UDP_GENEVE_OAM_CRI_OPT,
-		8, 0x3f, 0, 2,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP,
-		NPC_F_LD_L_UDP_GTP_GTPC,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 3, 0,
-		NPC_S_KPU12_TU_IP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GTPU,
-		NPC_F_LD_L_UDP_GTP_GTPU_G_PDU,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LD, NPC_LT_LD_UDP_GTPU,
-		NPC_F_LD_L_UDP_GTP_GTPU_UNK,
 		0, 0, 0, 0,
 	},
 	{
@@ -10532,193 +10104,193 @@ static struct npc_kpu_profile_action kpu8_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 4, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 4, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM,
+		NPC_F_LD_L_GRE_HAS_CSUM,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY,
+		NPC_F_LD_L_GRE_HAS_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_SEQ,
+		NPC_F_LD_L_GRE_HAS_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 16, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 16, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 4, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 4, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM,
+		NPC_F_LD_L_GRE_HAS_CSUM,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY,
+		NPC_F_LD_L_GRE_HAS_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 8, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_SEQ,
+		NPC_F_LD_L_GRE_HAS_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 12, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 6, 10, 0, 0,
-		NPC_S_KPU9_TU_MPLS_IN_GRE_VXLAN, 16, 1,
+		NPC_S_KPU9_TU_MPLS_IN_GRE, 16, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_MPLS_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 4, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 4, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 8, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM,
+		NPC_F_LD_L_GRE_HAS_CSUM,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 8, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY,
+		NPC_F_LD_L_GRE_HAS_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 8, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 8, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_SEQ,
+		NPC_F_LD_L_GRE_HAS_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 12, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 12, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 12, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 12, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
-		NPC_S_KPU9_TU_NSH, 16, 1,
+		NPC_S_KPU9_TU_NSH_IN_GRE, 16, 1,
 		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_U_NSH_IN_GRE|NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
+		NPC_F_LD_L_GRE_HAS_CSUM_KEY_SEQ,
 		0, 0, 0, 0,
 	},
 	{
@@ -10875,172 +10447,436 @@ static struct npc_kpu_profile_action kpu8_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU11_TU_PPP, 8, 1,
-		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_L_GRE_VER1,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU11_TU_PPP, 8, 1,
+		NPC_LID_LD, NPC_LT_LD_GRE,
+		NPC_F_LD_L_GRE_VER1,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU11_TU_PPP, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_GRE,
+		NPC_F_LD_L_GRE_VER1_HAS_SEQ,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU11_TU_PPP, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_GRE,
+		NPC_F_LD_L_GRE_VER1_HAS_ACK,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 2, 0,
+		NPC_S_KPU11_TU_PPP, 16, 1,
+		NPC_LID_LD, NPC_LT_LD_GRE,
+		NPC_F_LD_L_GRE_VER1_HAS_SEQ_ACK,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LD, NPC_LT_LD_GRE,
+		NPC_F_LD_L_GRE_VER1_UNK_PROTO,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LD, NPC_EC_GRE_VER1,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LD, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LD, NPC_EC_UNK,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LD, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+};
+
+static struct npc_kpu_profile_action kpu9_action_entries[] = {
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 4, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_MPLS_IN_GRE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_MPLS_IN_GRE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 12, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_MPLS_IN_GRE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 4, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS, 12, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_MPLS_IN_GRE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 4, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_NSH,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 8, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_NSH,
+		NPC_F_LD_L_MPLS_2_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_NSH,
+		NPC_F_LD_L_MPLS_3_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 4, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_NSH,
+		NPC_F_LD_L_MPLS_4_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 4, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_IP,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 8, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_IP,
+		NPC_F_LD_L_MPLS_2_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS_PL, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_IP,
+		NPC_F_LD_L_MPLS_3_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 4, 0, 0, 0,
+		NPC_S_KPU10_TU_MPLS, 12, 1,
+		NPC_LID_LD, NPC_LT_LD_TU_MPLS_IN_IP,
+		NPC_F_LD_L_MPLS_4_LABELS,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_NSH_IN_GRE,
+		0,
+		1, 0x3f, 0, 2,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_NSH_IN_GRE,
+		0,
+		1, 0x3f, 0, 2,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_NSH_IN_GRE,
+		0,
+		1, 0x3f, 0, 2,
+	},
+	{
+		NPC_ERRLEV_LE, NPC_EC_NSH_UNK,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_TU_NSH_IN_GRE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLAN,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLAN,
+		NPC_F_LE_L_VXLAN_NOVNI,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LE, NPC_EC_VXLAN,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LE, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 6, 10, 0, 0,
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NOVNI,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NOVNI,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NOVNI,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU11_TU_PPP, 12, 1,
-		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_L_GRE_VER1_HAS_SEQ,
+		2, 0, 0, 0, 0,
+		NPC_S_KPU10_TU_NSH_IN_VXLANGPE, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NOVNI,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU11_TU_PPP, 12, 1,
-		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_L_GRE_VER1_HAS_ACK,
+		2, 6, 10, 0, 0,
+		NPC_S_KPU10_TU_MPLS_IN_VXLANGPE, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NOVNI,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 2, 0,
-		NPC_S_KPU11_TU_PPP, 16, 1,
-		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_L_GRE_VER1_HAS_SEQ_ACK,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_UNK,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LD, NPC_LT_LD_GRE,
-		NPC_F_LD_L_GRE_VER1_UNK_PROTO,
+		NPC_LID_LE, NPC_LT_LE_VXLANGPE,
+		NPC_F_LE_L_VXLANGPE_NONP,
 		0, 0, 0, 0,
 	},
 	{
-		NPC_ERRLEV_LD, NPC_EC_GRE_VER1,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
 		0,
-		0, 0, 0, 0,
+		0, 0x3f, 0, 2,
 	},
 	{
-		NPC_ERRLEV_LD, NPC_EC_UNK,
-		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		0, 0, 0, 0,
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM,
+		0, 0x3f, 0, 2,
 	},
-};
-
-static struct npc_kpu_profile_action kpu9_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 4, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		0, 0, 0, 0,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 8, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		0, 0, 0, 0,
+		12, 16, 20, 1, 0,
+		NPC_S_KPU11_TU_ETHER, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 12, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
 		0,
-		0, 0, 0, 0,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 4, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS, 12, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		0, 0, 0, 0,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 4, 1,
-		NPC_LID_LD, NPC_LT_LD_TU_MPLS,
-		0,
-		0, 0, 0, 0,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 8, 1,
-		NPC_LID_LD, NPC_LT_LD_TU_MPLS,
-		NPC_F_LD_L_MPLS_2_LABELS,
-		0, 0, 0, 0,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS_PL, 12, 1,
-		NPC_LID_LD, NPC_LT_LD_TU_MPLS,
-		NPC_F_LD_L_MPLS_3_LABELS,
-		0, 0, 0, 0,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		0,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 4, 0, 0, 0,
-		NPC_S_KPU10_TU_MPLS, 12, 1,
-		NPC_LID_LD, NPC_LT_LD_TU_MPLS,
-		NPC_F_LD_L_MPLS_4_LABELS,
-		0, 0, 0, 0,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 2, 0,
-		NPC_S_KPU12_TU_IP, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
+		6, 0, 0, 2, 0,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 2, 0,
-		NPC_S_KPU12_TU_IP6, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
+		NPC_S_KPU12_TU_IP6, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GENEVE,
+		NPC_F_LE_L_GENEVE_OAM_CRI_OPT,
+		0, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 1, 0,
-		NPC_S_KPU11_TU_ETHER, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_GTPC,
 		0,
-		1, 0x3f, 0, 2,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 0, 0,
-		NPC_S_KPU10_TU_NSH, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
+		8, 0, 6, 2, 0,
+		NPC_S_KPU12_TU_IP, 8, 1,
+		NPC_LID_LE, NPC_LT_LE_GTPU,
+		NPC_F_LE_L_GTPU_G_PDU,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 1, 0,
-		NPC_S_KPU11_TU_MPLS_IN_NSH, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
-		0,
-		1, 0x3f, 0, 2,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LE, NPC_LT_LE_GTPU,
+		NPC_F_LE_L_GTPU_UNK,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_LE, NPC_EC_UNK,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LE, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11051,7 +10887,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 1, 0,
 		NPC_S_KPU12_TU_IP, 4, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11059,7 +10895,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 1, 0,
 		NPC_S_KPU12_TU_IP6, 4, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11067,7 +10903,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		12, 16, 20, 0, 0,
 		NPC_S_KPU11_TU_ETHER, 8, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11075,15 +10911,15 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		12, 16, 20, 0, 0,
 		NPC_S_KPU11_TU_ETHER, 4, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
-		NPC_ERRLEV_LB, NPC_EC_L2_MPLS_2MANY,
+		NPC_ERRLEV_LE, NPC_EC_MPLS_2MANY,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11091,7 +10927,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 1, 0,
 		NPC_S_KPU12_TU_IP, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11099,7 +10935,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 1, 0,
 		NPC_S_KPU12_TU_IP6, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11107,7 +10943,7 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		12, 16, 20, 0, 0,
 		NPC_S_KPU11_TU_ETHER, 4, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11115,55 +10951,79 @@ static struct npc_kpu_profile_action kpu10_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		12, 16, 20, 0, 0,
 		NPC_S_KPU11_TU_ETHER, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 0, 6, 1, 0,
-		NPC_S_KPU12_TU_IP, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU11_TU_MPLS_PL, 4, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_MPLS_IN_VXLANGPE,
 		0,
-		1, 0x3f, 0, 2,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		6, 0, 0, 1, 0,
-		NPC_S_KPU12_TU_IP6, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU11_TU_MPLS_PL, 8, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_MPLS_IN_VXLANGPE,
 		0,
-		1, 0x3f, 0, 2,
+		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		12, 16, 20, 0, 0,
-		NPC_S_KPU11_TU_ETHER, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		0, 0, 0, 0, 0,
+		NPC_S_KPU11_TU_MPLS_PL, 12, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_MPLS_IN_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		2, 4, 0, 0, 0,
+		NPC_S_KPU11_TU_MPLS, 12, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_MPLS_IN_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 1, 0,
+		NPC_S_KPU12_TU_IP, 0, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_NSH_IN_VXLANGPE,
 		0,
 		1, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 0, 0,
-		NPC_S_KPU11_TU_3RD_NSH, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		6, 0, 0, 1, 0,
+		NPC_S_KPU12_TU_IP6, 0, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_NSH_IN_VXLANGPE,
 		0,
 		1, 0x3f, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 0, 0,
-		NPC_S_KPU11_TU_MPLS_IN_NSH, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		6, 0, 0, 0, 0,
+		NPC_S_KPU11_TU_ETHER_IN_NSH, 0, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_NSH_IN_VXLANGPE,
 		0,
 		1, 0x3f, 0, 2,
 	},
+	{
+		NPC_ERRLEV_LF, NPC_EC_NSH_UNK,
+		6, 0, 0, 0, 1,
+		NPC_S_NA, 0, 1,
+		NPC_LID_LF, NPC_LT_LF_TU_NSH_IN_VXLANGPE,
+		0,
+		0, 0, 0, 0,
+	},
 	{
 		NPC_ERRLEV_LE, NPC_EC_UNK,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LD, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11174,7 +11034,7 @@ static struct npc_kpu_profile_action kpu11_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 14, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11182,7 +11042,7 @@ static struct npc_kpu_profile_action kpu11_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 14, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11190,7 +11050,7 @@ static struct npc_kpu_profile_action kpu11_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 14, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11198,183 +11058,231 @@ static struct npc_kpu_profile_action kpu11_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE|NPC_F_LE_L_WITH_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE | NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_STAG_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_STAG_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_STAG_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE|NPC_F_LE_L_WITH_STAG_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE | NPC_F_LF_L_WITH_STAG_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_STAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE|NPC_F_LE_L_WITH_STAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE | NPC_F_LF_L_WITH_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 22, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE|NPC_F_LE_L_WITH_QINQ_CTAG,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE | NPC_F_LF_L_WITH_QINQ_CTAG,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		8, 0, 6, 0, 0,
 		NPC_S_KPU12_TU_IP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		6, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_IP6, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU12_TU_ARP, 18, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_L_WITH_QINQ,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_L_WITH_QINQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE|NPC_F_LE_L_WITH_QINQ,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE | NPC_F_LF_L_WITH_QINQ,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_ETHER,
-		NPC_F_LE_U_UNK_ETYPE,
+		NPC_LID_LF, NPC_LT_LF_TU_ETHER,
+		NPC_F_LF_U_UNK_ETYPE,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_PPP,
+		NPC_LID_LF, NPC_LT_LF_TU_PPP,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 0, 0,
+		NPC_S_KPU12_TU_IP, 4, 0,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		6, 0, 0, 0, 0,
+		NPC_S_KPU12_TU_IP6, 4, 0,
+		NPC_LID_LF, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LF, NPC_EC_MPLS_UNK,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_MPLS_IN_NSH,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LF, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LF, NPC_EC_MPLS_2MANY,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LF, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		8, 0, 6, 0, 0,
+		NPC_S_KPU12_TU_IP, 0, 0,
+		NPC_LID_LF, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_RE, NPC_EC_NOERR,
+		6, 0, 0, 0, 0,
+		NPC_S_KPU12_TU_IP6, 0, 0,
+		NPC_LID_LF, NPC_LT_NA,
+		0,
+		0, 0, 0, 0,
+	},
+	{
+		NPC_ERRLEV_LF, NPC_EC_MPLS_UNK,
+		0, 0, 0, 0, 1,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11382,15 +11290,15 @@ static struct npc_kpu_profile_action kpu11_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LE, NPC_LT_LE_TU_3RD_NSH,
+		NPC_LID_LG, NPC_LT_LG_TU_ETHER_IN_NSH,
 		0,
 		0, 0, 0, 0,
 	},
 	{
-		NPC_ERRLEV_LE, NPC_EC_UNK,
+		NPC_ERRLEV_LF, NPC_EC_UNK,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LE, NPC_LT_NA,
+		NPC_LID_LF, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11401,7 +11309,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 12, 0, 2, 0,
 		NPC_S_KPU15_TU_TCP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11409,7 +11317,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_UDP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11417,7 +11325,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_SCTP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11425,7 +11333,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ICMP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11433,7 +11341,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_IGMP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11441,7 +11349,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ESP, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11449,7 +11357,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_AH, 20, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11457,79 +11365,79 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_UNK_IP_PROTO,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_UNK_IP_PROTO,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 12, 0, 2, 0,
 		NPC_S_KPU15_TU_TCP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_UDP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_SCTP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ICMP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_IGMP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ESP, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_AH, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS,
 		0, 0xf, 0, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
-		NPC_F_LF_U_IP_HAS_OPTIONS|NPC_F_LF_U_UNK_IP_PROTO,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
+		NPC_F_LG_U_IP_HAS_OPTIONS | NPC_F_LG_U_UNK_IP_PROTO,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_LF, NPC_EC_IP_VER,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP,
+		NPC_LID_LG, NPC_LT_LG_TU_IP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11537,7 +11445,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_ARP,
+		NPC_LID_LG, NPC_LT_LG_TU_ARP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11545,7 +11453,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 12, 0, 2, 0,
 		NPC_S_KPU15_TU_TCP, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11553,7 +11461,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_UDP, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11561,7 +11469,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_SCTP, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11569,7 +11477,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ICMP, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11577,7 +11485,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ICMP6, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11585,7 +11493,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_ESP, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11593,7 +11501,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 2, 0,
 		NPC_S_KPU15_TU_AH, 40, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11601,15 +11509,15 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		2, 0, 0, 0, 0,
 		NPC_S_KPU13_TU_IP6_EXT, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
-		NPC_F_LF_U_IP6_HAS_EXT,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
+		NPC_F_LG_U_IP6_HAS_EXT,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_LF, NPC_EC_IP6_VER,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LF, NPC_LT_LF_TU_IP6,
+		NPC_LID_LG, NPC_LT_LG_TU_IP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11617,7 +11525,7 @@ static struct npc_kpu_profile_action kpu12_action_entries[] = {
 		NPC_ERRLEV_LF, NPC_EC_UNK,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LF, NPC_LT_NA,
+		NPC_LID_LG, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11650,7 +11558,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_FIN_ONLY,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11658,7 +11566,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_ZERO,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11666,7 +11574,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_RST_FIN,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11674,7 +11582,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_URG_SYN,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11682,7 +11590,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_RST_SYN,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11690,7 +11598,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_TCP_FLAGS_SYN_FIN,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11698,79 +11606,79 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_HTTP_DATA, 20, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_L_TCP_HTTP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_L_TCP_HTTP,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_HTTPS_DATA, 20, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_L_TCP_HTTPS,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_L_TCP_HTTP,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_PPTP_DATA, 20, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_L_TCP_PPTP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_L_TCP_PPTP,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_TCP_DATA, 20, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_L_TCP_UNK_PORT,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_L_TCP_UNK_PORT,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_HTTP_DATA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_U_TCP_HAS_OPTIONS|NPC_F_LG_L_TCP_HTTP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_U_TCP_HAS_OPTIONS | NPC_F_LH_L_TCP_HTTP,
 		12, 0xf0, 1, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_HTTPS_DATA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_U_TCP_HAS_OPTIONS|NPC_F_LG_L_TCP_HTTPS,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_U_TCP_HAS_OPTIONS | NPC_F_LH_L_TCP_HTTPS,
 		12, 0xf0, 1, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_PPTP_DATA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_U_TCP_HAS_OPTIONS|NPC_F_LG_L_TCP_PPTP,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_U_TCP_HAS_OPTIONS | NPC_F_LH_L_TCP_PPTP,
 		12, 0xf0, 1, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_TCP_DATA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_TCP,
-		NPC_F_LG_U_TCP_HAS_OPTIONS|NPC_F_LG_L_TCP_UNK_PORT,
+		NPC_LID_LH, NPC_LT_LH_TU_TCP,
+		NPC_F_LH_U_TCP_HAS_OPTIONS | NPC_F_LH_L_TCP_UNK_PORT,
 		12, 0xf0, 1, 2,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 0,
 		NPC_S_KPU16_UDP_DATA, 8, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_UDP,
-		NPC_F_LG_L_UDP_UNK_PORT,
+		NPC_LID_LH, NPC_LT_LH_TU_UDP,
+		NPC_F_LH_L_UDP_UNK_PORT,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_SCTP,
+		NPC_LID_LH, NPC_LT_LH_TU_SCTP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11778,7 +11686,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_ICMP,
+		NPC_LID_LH, NPC_LT_LH_TU_ICMP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11786,7 +11694,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_IGMP,
+		NPC_LID_LH, NPC_LT_LH_TU_IGMP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11794,7 +11702,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_ICMP6,
+		NPC_LID_LH, NPC_LT_LH_TU_ICMP6,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11802,7 +11710,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_ESP,
+		NPC_LID_LH, NPC_LT_LH_TU_ESP,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11810,7 +11718,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 1,
-		NPC_LID_LG, NPC_LT_LG_TU_AH,
+		NPC_LID_LH, NPC_LT_LH_TU_AH,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11818,7 +11726,7 @@ static struct npc_kpu_profile_action kpu15_action_entries[] = {
 		NPC_ERRLEV_LG, NPC_EC_L4,
 		0, 0, 0, 0, 1,
 		NPC_S_NA, 0, 0,
-		NPC_LID_LG, NPC_LT_NA,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
@@ -11828,48 +11736,48 @@ static struct npc_kpu_profile_action kpu16_action_entries[] = {
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_TCP_DATA,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_HTTP_DATA,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_HTTPS_DATA,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_PPTP_DATA,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_UDP_DATA,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		0, 0, 0, 0, 1,
-		NPC_S_NA, 0, 1,
-		NPC_LID_LH, NPC_LT_LH_UDP_PTP,
+		NPC_S_NA, 0, 0,
+		NPC_LID_LH, NPC_LT_NA,
 		0,
 		0, 0, 0, 0,
 	},
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
index abcea17cb12d..353694a0ab2d 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_nix.c
@@ -2587,8 +2587,8 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			field->lid = NPC_LID_LC;
 			field->ltype_match = NPC_LT_LC_IP;
 			if (key_type == NIX_FLOW_KEY_TYPE_INNR_IPV4) {
-				field->lid = NPC_LID_LF;
-				field->ltype_match = NPC_LT_LF_TU_IP;
+				field->lid = NPC_LID_LG;
+				field->ltype_match = NPC_LT_LG_TU_IP;
 			}
 			field->hdr_offset = 12; /* SIP offset */
 			field->bytesm1 = 7; /* SIP + DIP, 8 bytes */
@@ -2600,8 +2600,8 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			field->lid = NPC_LID_LC;
 			field->ltype_match = NPC_LT_LC_IP6;
 			if (key_type == NIX_FLOW_KEY_TYPE_INNR_IPV6) {
-				field->lid = NPC_LID_LF;
-				field->ltype_match = NPC_LT_LF_TU_IP6;
+				field->lid = NPC_LID_LG;
+				field->ltype_match = NPC_LT_LG_TU_IP6;
 			}
 			field->hdr_offset = 8; /* SIP offset */
 			field->bytesm1 = 31; /* SIP + DIP, 32 bytes */
@@ -2625,11 +2625,11 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			 * the lid for inner protocols
 			 */
 			BUILD_BUG_ON((int)NPC_LT_LD_TCP !=
-				     (int)NPC_LT_LG_TU_TCP);
+				     (int)NPC_LT_LH_TU_TCP);
 			BUILD_BUG_ON((int)NPC_LT_LD_UDP !=
-				     (int)NPC_LT_LG_TU_UDP);
+				     (int)NPC_LT_LH_TU_UDP);
 			BUILD_BUG_ON((int)NPC_LT_LD_SCTP !=
-				     (int)NPC_LT_LG_TU_SCTP);
+				     (int)NPC_LT_LH_TU_SCTP);
 
 			if ((key_type == NIX_FLOW_KEY_TYPE_TCP ||
 			     key_type == NIX_FLOW_KEY_TYPE_INNR_TCP) &&
@@ -2663,27 +2663,29 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			}
 			break;
 		case NIX_FLOW_KEY_TYPE_NVGRE:
-		case NIX_FLOW_KEY_TYPE_VXLAN:
-		case NIX_FLOW_KEY_TYPE_GENEVE:
 			field->lid = NPC_LID_LD;
 			field->bytesm1 = 2;
 			field->ltype_mask = 0xF;
 			keyoff_marker = false;
-			if (key_type == NIX_FLOW_KEY_TYPE_NVGRE && valid_key) {
+			if (valid_key) {
 				field->hdr_offset = 4; /* VSID offset */
 				field->ltype_match = NPC_LT_LD_GRE;
 			}
-
+			break;
+		case NIX_FLOW_KEY_TYPE_VXLAN:
+		case NIX_FLOW_KEY_TYPE_GENEVE:
+			field->lid = NPC_LID_LE;
+			field->bytesm1 = 2;
+			field->ltype_mask = 0xF;
+			keyoff_marker = false;
 			if (key_type == NIX_FLOW_KEY_TYPE_VXLAN && valid_key) {
-				/* VNI at UDP header + 4B */
-				field->hdr_offset = 12;
-				field->ltype_match = NPC_LT_LD_UDP_VXLAN;
+				field->hdr_offset = 4;
+				field->ltype_match = NPC_LT_LE_VXLAN;
 			}
 
 			if (key_type == NIX_FLOW_KEY_TYPE_GENEVE && valid_key) {
-				/* VNI at UDP header + 4B */
-				field->hdr_offset = 12;
-				field->ltype_match = NPC_LT_LD_UDP_GENEVE;
+				field->hdr_offset = 4;
+				field->ltype_match = NPC_LT_LE_GENEVE;
 			}
 
 			if (key_type == NIX_FLOW_KEY_TYPE_GENEVE)
@@ -2694,8 +2696,8 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			field->lid = NPC_LID_LA;
 			field->ltype_match = NPC_LT_LA_ETHER;
 			if (key_type == NIX_FLOW_KEY_TYPE_INNR_ETH_DMAC) {
-				field->lid = NPC_LID_LE;
-				field->ltype_match = NPC_LT_LE_TU_ETHER;
+				field->lid = NPC_LID_LF;
+				field->ltype_match = NPC_LT_LF_TU_ETHER;
 			}
 			field->hdr_offset = 0;
 			field->bytesm1 = 5; /* DMAC 6 Byte */
@@ -2709,33 +2711,33 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			field->ltype_mask = 0xF;
 			break;
 		case NIX_FLOW_KEY_TYPE_GTPU:
-			field->lid = NPC_LID_LD;
-			field->hdr_offset = 12; /* UDP + hdr */
+			field->lid = NPC_LID_LE;
+			field->hdr_offset = 0; /* UDP data */
 			field->bytesm1 = 3; /* 4 bytes VNI*/
-			field->ltype_match = NPC_LT_LD_UDP_GTPU;
+			field->ltype_match = NPC_LT_LE_GTPU;
 			field->ltype_mask = 0xF;
 			break;
 		case NIX_FLOW_KEY_TYPE_UDP_VXLAN:
 		case NIX_FLOW_KEY_TYPE_UDP_GENEVE:
-			field->lid = NPC_LID_LD;
+			field->lid = NPC_LID_LE;
 			field->hdr_offset = 0; /* UDP data */
 			field->bytesm1 = 3; /* 4 bytes SPORT +  DPORT*/
 			field->ltype_mask = 0xF;
 			field_marker = false;
 			keyoff_marker = false;
-			BUILD_BUG_ON(NPC_LT_LD_UDP_VXLAN != 12);
-			BUILD_BUG_ON(NPC_LT_LD_UDP_GENEVE != 13);
+			BUILD_BUG_ON(NPC_LT_LE_VXLAN != 1);
+			BUILD_BUG_ON(NPC_LT_LE_GENEVE != 2);
 			/* Only VXLAN enabled */
 			if (key_type == NIX_FLOW_KEY_TYPE_UDP_VXLAN &&
 			    valid_key) {
-				field->ltype_match |= NPC_LT_LD_UDP_VXLAN;
+				field->ltype_match |= NPC_LT_LE_VXLAN;
 				udp_tu_data |= (1 << 0);
 			}
 
 			/* Only GENEVE enabled */
 			if (key_type == NIX_FLOW_KEY_TYPE_UDP_GENEVE &&
 			    valid_key) {
-				field->ltype_match |= NPC_LT_LD_UDP_GENEVE;
+				field->ltype_match |= NPC_LT_LE_GENEVE;
 				udp_tu_data |= (1 << 1);
 			}
 
@@ -2755,7 +2757,7 @@ static int set_flowkey_fields(struct nix_rx_flowkey_alg *alg, u32 flow_cfg)
 			field->lid = NPC_LID_LD;
 			field->hdr_offset = 0; /* UDP data */
 			field->bytesm1 = 3; /* 4 bytes SPORT +  DPORT*/
-			field->ltype_match = NPC_LT_LD_UDP_GTPU;
+			field->ltype_match = NPC_LT_LD_UDP;
 			field->ltype_mask = 0xF;
 			break;
 		}
@@ -3393,23 +3395,23 @@ int rvu_nix_init(struct rvu *rvu)
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_OIP4,
 			    (NPC_LID_LC << 8) | (NPC_LT_LC_IP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_IIP4,
-			    (NPC_LID_LF << 8) | (NPC_LT_LF_TU_IP << 4) | 0x0F);
+			    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_IP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_OIP6,
 			    (NPC_LID_LC << 8) | (NPC_LT_LC_IP6 << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_IIP6,
-			    (NPC_LID_LF << 8) | (NPC_LT_LF_TU_IP6 << 4) | 0x0F);
+			    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_IP6 << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_OTCP,
 			    (NPC_LID_LD << 8) | (NPC_LT_LD_TCP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_ITCP,
-			    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_TCP << 4) | 0x0F);
+			    (NPC_LID_LH << 8) | (NPC_LT_LH_TU_TCP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_OUDP,
 			    (NPC_LID_LD << 8) | (NPC_LT_LD_UDP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_IUDP,
-			    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_UDP << 4) | 0x0F);
+			    (NPC_LID_LH << 8) | (NPC_LT_LH_TU_UDP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_OSCTP,
 			    (NPC_LID_LD << 8) | (NPC_LT_LD_SCTP << 4) | 0x0F);
 		rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_ISCTP,
-			    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_SCTP << 4) |
+			    (NPC_LID_LH << 8) | (NPC_LT_LH_TU_SCTP << 4) |
 			    0x0F);
 
 		err = nix_rx_flowkey_alg_cfg(rvu, blkaddr);
@@ -3433,8 +3435,8 @@ int rvu_nix_init(struct rvu *rvu)
 				    (NPC_LT_LD_ESP << 4) | 0x0F);
 
 			rvu_write64(rvu, blkaddr, NIX_AF_RX_DEF_IPSECX(1),
-				    (8 << 12) | (NPC_LID_LG << 8) |
-				    (NPC_LT_LG_TU_ESP << 4) | 0x0F);
+				    (8 << 12) | (NPC_LID_LH << 8) |
+				    (NPC_LT_LH_TU_ESP << 4) | 0x0F);
 		}
 	}
 	return 0;
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
index b8d51ef7cebe..e0be2a1a4476 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc.c
@@ -1357,7 +1357,7 @@ int rvu_npc_init(struct rvu *rvu)
 
 	/* Config Inner IPV4 NPC layer info */
 	rvu_write64(rvu, blkaddr, NPC_AF_PCK_DEF_IIP4,
-		    (NPC_LID_LF << 8) | (NPC_LT_LF_TU_IP << 4) | 0x0F);
+		    (NPC_LID_LG << 8) | (NPC_LT_LG_TU_IP << 4) | 0x0F);
 
 	/* Enable below for Rx pkts.
 	 * - Outer IPv4 header checksum validation.
-- 
2.31.1

